### PR TITLE
Switch to src layout, and some minor modifications.

### LINKED ISSUE #6: Proposal: switch to src layout instead of flat/ad-hoc layout
I'd like to switch the layout of the project template over to the src layout. The two types of layouts I'm referencing are [these](https://packaging.python.org/en/latest/discussions/src-layout-vs-flat-layout/).

The reason I'd like this change is that, as hinted at in the breakdown in the link above, the flat layout can lead to some subtle error when testing or configuring package distribution. This is due to the fact that Python interpreter includes the current working directory as the first item on the import path, so e.g. when running tests they get run against the current state of the repo, and not against what a user would get if they installed the package using pip. These links have some further examples: [link1](https://stackoverflow.com/questions/72712965/does-the-src-folder-in-pypi-packaging-have-a-special-meaning-or-is-it-only-a-co) [link2](https://hynek.me/articles/testing-packaging/).

To appeal to a higher authority, the [pypa packaging tutorial](https://packaging.python.org/en/latest/tutorials/packaging-projects/) also uses the src layout due to these points;)

I'd be happy to open a pull request if we agree that this is something we'd like to do. We could also include some instructions on how to create an editable install and how to build a project for beginners to reduce any possible friction. Let me know your thoughts @rkaminsk @susuhahnml!

### PR DESCRIPTION
**Pull request by namcsi**
Resolves #6

### CODE REVIEW

File: setup.cfg
Comment (rkaminsk) [inline setup.cfg:17]: The template uses `pkg_resources` so `setuptools` is actually a hard requirement. I don't know if listing it in `pyproject.toml` is sufficient.

We should probably switch to `import_metadata`. See <https://setuptools.pypa.io/en/latest/pkg_resources.html>. Currently, the template explicitly supports Python 3.7. I always try to support Python versions that are still maintained (and even a bit beyond that). There also seems to be a backport of the library.

File: setup.cfg
Comment (rkaminsk) [inline setup.cfg:15]: What does the `=` sign before src mean?

File: src/fillname/__init__.py
Comment (rkaminsk) [inline src/fillname/__init__.py:1]: Is there a reason to overwrite `__init__.py` with `__main__.py`. There might be a lot of unnecessary imports in case the project is intend to be usable as a library too.

**Review by rkaminsk — CHANGES_REQUESTED** [review]
We still have to discuss testing and editable installs + some minor things. Also CI tests have to work before merging anything.

File: README.md
Comment (namcsi) [reply README.md:1]: The main motivation was this disclaimer on the [editable install setuptools page](https://setuptools.pypa.io/en/latest/userguide/development_mode.html#limitations):

"Editable installs are not a perfect replacement for regular installs in a test environment. When in doubt, please test your projects as installed via a regular wheel. There are tools in the Python ecosystem, like [tox](https://pypi.org/project/tox) or [nox](https://pypi.org/project/nox), that can help you with that (when used with appropriate configuration)."

I also found that the editable install thing gets a bit annoying when running tests locally as well, as non python files don't get auto-updated, so testing after editing my e.g. my encoding files can cause inaccurate test results. Though this can be fixed by a quick re-install of the environment, it might trip up new users:)

Probably the benefit for fast testing would be worth it for most people though so we could just provide a separate test session that doesn't use an editable install, and have the CI run that session.

File: setup.cfg
Comment (namcsi) [reply setup.cfg:15]: It's some syntactic sugar. It indicates that all packages are contained under the directory. It's described [here](https://setuptools.pypa.io/en/latest/userguide/declarative_config.html#using-a-src-layout).

File: src/fillname/__init__.py
Comment (namcsi) [reply src/fillname/__init__.py:1]: Yes, I switched things up because the console_scrips entry point wasn't working.
I've since figured out a way to edit the setup.cfg entry make it work with the main function defined in __main__.py so I can move it back.

Comment (namcsi) [reply src/fillname/__init__.py:1]: Done in 80af5c168141a4593080887983880d63161c7c55

File: init.py
Comment (namcsi) [reply init.py]: Done in 5f1f04d2b7f79e67a71ba2f840cb66f58d27a302

File: init.py
Comment (namcsi) [reply init.py]: Done in 5f1f04d2b7f79e67a71ba2f840cb66f58d27a302.

**Comment by namcsi** [comment]
CI passes now. I think the previous failures were due to a Github outage.

File: setup.cfg
Comment (namcsi) [reply setup.cfg:17]: I've tested this, and it's sufficient to `setuptools` it in `pyproject.toml`, `pkg_resources` is importable, and the version number shows up on the command line as expected. But since it's deprecated I can put together a commit switching to using `importlib.metadata` / `importlib_metadata` (depending on the python version) to check the version number.

File: README.md
Comment (rkaminsk) [reply README.md:1]: > The main motivation was this disclaimer on the [editable install setuptools page](https://setuptools.pypa.io/en/latest/userguide/development_mode.html#limitations):
> 
> "Editable installs are not a perfect replacement for regular installs in a test environment. When in doubt, please test your projects as installed via a regular wheel. There are tools in the Python ecosystem, like [tox](https://pypi.org/project/tox) or [nox](https://pypi.org/project/nox), that can help you with that (when used with appropriate configuration)."
> 
> I also found that testing in the editable install thing gets a bit annoying when running tests locally as well, as non python files don't get auto-updated, so testing after editing my e.g. my encoding files can cause inaccurate test results. Though this can be fixed by a quick re-install of the environment, it might trip up new users:)

I don't see so much of an issue. By default, nox recreates the environment anyway. 

> Probably the benefit of fast testing would be worth it for most people though so we could just provide a separate test session that doesn't use an editable install, and have the CI run that session. I can add a commit doing so and update the github workflow file.

I rather have the default test session and add parameters to change the default behavior. This should be easy with nox. If it is better for a project to disable editable installs altogether, then they can simply be disabled. This is just a template after all.

File: src/fillname/utils/parser.py
Comment (rkaminsk) [inline src/fillname/utils/parser.py:18]: Since the source layout kind of forces installation, we can probably avoid catching `metadata.PackageNotFoundError`.

Comment (namcsi) [reply src/fillname/utils/parser.py:18]: Agreed!

File: README.md
Comment (rkaminsk) [reply README.md:1]: See the last two commits trying to address this.

File: .envrc
Comment (namcsi) [reply .envrc:1]: Done in [2184111](https://github.com/krr-up/python-project-template/pull/7/commits/218411167ee6f5e59e7fe2e6b6b14425e5ff7fc8)

File: README.md
Comment (namcsi) [reply README.md:1]: Looks good to me, working as expected in the CI runs!

**Comment by namcsi** [comment]
I went through the creation of a new package based on the template, and after fixing a small error in init.py and tweaking .coveragerc to collect metrics from the execution of the installed package, everything looks to be working as intended now. I'd be comfortable merging now from my side, let me know if you want to tweak the README.md or other stuff @rkaminsk. I can do an interactive rebase and squash some commits to clean up the commit history a bit before merging if you like.

**Comment by rkaminsk** [comment]
> I went through the creation of a new package based on the template, and after fixing a small error in init.py and tweaking .coveragerc to collect metrics correctly, everything looks to be working as intended now. 

:+1: 

> I'd be comfortable merging now from my side, let me know if you want to tweak the README.md or other stuff @rkaminsk.

I'll do another pass over the PR later and also try to actually use the template. Probably won't need more changes from you.

>  I can do an interactive rebase and squash some commits to clean up the commit history a bit before merging if you like.

I personally don't care because I don't think this is important here. Do as you like.

**Comment by namcsi** [comment]
> I'll do another pass over the PR later and also try to actually use the template. Probably won't need more changes from you.

Alright then, I leave it up to you!

> I personally don't care because I don't think this is important here. Do as you like.

Same, I won't bother in that case.

**Pull request merged**
